### ISSUE TITLE
Feature Request: git verify-commit

### ISSUE THREAD

**Issue by mohamedhafez** [capistrano/capistrano#2074]
For those of us that GPG sign our git commits, it would be great to have Capistrano make sure that when we deploy, the commit we are deploying has been signed by a trusted person. (e.g., to prevent against deploying code [in a case like this](https://github.blog/2012-03-04-public-key-security-vulnerability-and-mitigation/))

I've just added the following line at the end of `Capistrano::SCM::Git#update_mirror` to do this:
```ruby
git(:"verify-commit", fetch_revision) if fetch(:verify_commit, false)
```

Let me know what you guys think!

**Comment by mohamedhafez** [comment]
the line above will accept any GPG signature for which you have the public key in your keyring, whether it's marked as trusted or not. I don't really know how to make it do that though aside from capturing the output and checking if it says "WARNING" or not, perhaps someone out there has a suggestion?

**Comment by leehambley** [comment]
Hi @mohamedhafez this is absolutely something you should be able to add by patching the rask tasks.

I'm not sure how familiar with Rake (Capistrano) you are, but you should be able to prepend a check, and define your own variables, and tie this in without requiring a patch in Capistrano itself.

I think we'd definitely take a patch for something like this, I can't quite gauge now whether it's more sensible to shell-out to GPG tooling, or find some Gem that binds, or reimplements PGP for checking the signatures. With us shelling out to Git anyway, maybe it makes sense to shell out to Git's signature verification tools.

If you check the flow of tasks here https://capistranorb.com/documentation/getting-started/flow/ then you can see where you might want to integrate this check by reopening the `task` and plugging in the checks you are seeing.

Please let me know if you need/want review on any WIP PRs, this should be fairly easy to add.

**Comment by mohamedhafez** [comment]
I was thinking that if we are using Git's signature verification tools, then it would make sense to have this be a part of `Capistrano::SCM::Git`, so we inherit all the defaults and the wrapper and environment variables, etc, a user might have set for Git. Although it would be easy to make the verification a separate task that hooked in after `git:update` if you guys think that would be cleaner for a PR.

**Comment by leehambley** [comment]
If you reopen `git:update` by just _defining again_ then you will be able to modify that task and we can definitely take a patch in the Git SCM plugin if/when you think it's ready to go.

**Comment by mohamedhafez** [comment]
Ah cool, didn't realize it worked that way! yeah i am a newbie to Capistrano, and just know the basics of Rake in general. 

I've been running with the above patch for about the past week and its been working perfectly, if I submit a PR I'm guessing you guys wouldn't want a 2nd definition of `git:update` in the PR though, right? Would you prefer a separate `git:verify` task that hooked in after git:update, or maybe just adding a line to the `git:update` task like

```ruby
git_plugin.verify_commit if fetch(:verify_commit)
```

and then defining a `verify_commit` method in `Capistrano::SCM::Git`?

Sorry, just not sure what the preferred style would be being new to this. Either would work, so just checking what you guys would prefer.

**Comment by mohamedhafez** [comment]
Or do you mean just send in a PR for a `verify_commit` method in `Capistrano::SCM::Git`, and leave it to users to hook it in if they wish?

**Comment by leehambley** [comment]
Haha, I'm only just realizing that `git(:"verify-commit", fetch_revision) if fetch(:verify_commit, false)` is the entirety of your patch! 

Got it, so yeah, you can send us this in a PR (if you need a hand writing tests/etc just let us know, they're _fairly_ intuitive). I think a complete patch to the git plugin might be something like this:

```
namespace :git do
  task :update do
     git(:"verify-commit", fetch_revision) if fetch(:verify_commit, false)
  end
end
```

(I don't have the code open right now, so I'm speculating a bit, and actually, this is slightly too late, as it will apply _after_ git update)

**Comment by williamdes** [comment]
This should be closed because https://github.com/capistrano/capistrano/pull/2076 fixed it

**Comment by mattbrictson** [comment]
Fixed in #2076
